File: server/controllers/booking_controller.py
from datetime import datetime
from flask import request
from config.db_config import SessionLocal, COMMISSION_RATE
from models.models import Booking, Vendor, Event, Payment
from services.payment_service import calculate_payment_breakdown
from services.refund_service import process_demo_cancellation_and_refund, calculate_cancellation_refund
from services.commission_service import calculate_commission
from utils.helpers import success_response, error_response, generate_booking_reference

def create_booking(current_user):
    data = request.get_json() or {}
    vendor_id = data.get('vendor_id')
    event_id = data.get('event_id')
    event_date_str = data.get('event_date')
    service_category = data.get('service_category', 'General Service')
    total_amount = float(data.get('total_amount', 0))
    special_notes = data.get('special_notes', '')

    if not vendor_id or not event_date_str or total_amount <= 0:
        return error_response("Vendor, valid event date, and total amount are required", 400)

    try:
        event_date = datetime.strptime(event_date_str, '%Y-%m-%d').date()
    except ValueError:
        return error_response("Invalid date format. Use YYYY-MM-DD", 400)

    db = SessionLocal()
    try:
        vendor = db.query(Vendor).filter(Vendor.id == vendor_id).first()
        if not vendor:
            return error_response("Vendor not found", 404)

        # Payment split: 20% advance, 80% remaining
        payment_split = calculate_payment_breakdown(total_amount)
        comm = calculate_commission(total_amount, COMMISSION_RATE)

        booking_ref = generate_booking_reference()

        new_booking = Booking(
            booking_reference=booking_ref,
            user_id=current_user.id,
            event_id=event_id,
            vendor_id=vendor_id,
            event_date=event_date,
            service_category=service_category or vendor.category,
            total_amount=total_amount,
            advance_amount=payment_split['advance_amount'],
            remaining_amount=payment_split['remaining_amount'],
            booking_status='PENDING',
            payment_status='PENDING',
            commission_rate=COMMISSION_RATE,
            commission_amount=comm['commission_amount'],
            special_notes=special_notes
        )
        db.add(new_booking)
        db.commit()
        db.refresh(new_booking)

        return success_response(new_booking.to_dict(), "Booking initiated successfully. Ready for advance payment.", 201)

    except Exception as e:
        db.rollback()
        return error_response(f"Booking creation failed: {str(e)}", 500)
    finally:
        db.close()

def get_bookings(current_user):
    db = SessionLocal()
    try:
        if current_user.role == 'VENDOR':
            # Find vendor profile for current user
            vendor = db.query(Vendor).filter(Vendor.user_id == current_user.id).first()
            if vendor:
                bookings = db.query(Booking).filter(Booking.vendor_id == vendor.id).order_by(Booking.created_at.desc()).all()
            else:
                bookings = []
        elif current_user.role == 'ADMIN':
            bookings = db.query(Booking).order_by(Booking.created_at.desc()).all()
        else:
            # CUSTOMER
            bookings = db.query(Booking).filter(Booking.user_id == current_user.id).order_by(Booking.created_at.desc()).all()

        return success_response({
            'count': len(bookings),
            'bookings': [b.to_dict() for b in bookings]
        }, "Bookings retrieved", 200)
    finally:
        db.close()

def get_booking_by_id(booking_id, current_user):
    db = SessionLocal()
    try:
        booking = db.query(Booking).filter(Booking.id == booking_id).first()
        if not booking:
            return error_response("Booking not found", 404)

        # Check authorization
        is_owner = (booking.user_id == current_user.id)
        is_vendor = (booking.vendor and booking.vendor.user_id == current_user.id)
        is_admin = (current_user.role == 'ADMIN')

        if not (is_owner or is_vendor or is_admin):
            return error_response("Unauthorized access to this booking", 403)

        booking_dict = booking.to_dict()
        booking_dict['cancellation_policy_preview'] = calculate_cancellation_refund(booking)
        return success_response(booking_dict, "Booking details retrieved", 200)
    finally:
        db.close()

def update_booking_status(booking_id, current_user):
    data = request.get_json() or {}
    new_status = data.get('booking_status')
    
    if not new_status:
        return error_response("New booking status is required", 400)

    db = SessionLocal()
    try:
        booking = db.query(Booking).filter(Booking.id == booking_id).first()
        if not booking:
            return error_response("Booking not found", 404)

        booking.booking_status = new_status
        db.commit()
        db.refresh(booking)
        return success_response(booking.to_dict(), f"Booking status updated to {new_status}", 200)
    except Exception as e:
        db.rollback()
        return error_response(f"Failed to update booking status: {str(e)}", 500)
    finally:
        db.close()

def cancel_booking(booking_id, current_user):
    data = request.get_json() or {}
    reason = data.get('reason', 'Customer requested cancellation')

    db = SessionLocal()
    try:
        booking = db.query(Booking).filter(Booking.id == booking_id).first()
        if not booking:
            return error_response("Booking not found", 404)

        is_owner = (booking.user_id == current_user.id)
        is_vendor = (booking.vendor and booking.vendor.user_id == current_user.id)
        is_admin = (current_user.role == 'ADMIN')

        if not (is_owner or is_vendor or is_admin):
            return error_response("Unauthorized to cancel this booking", 403)

        if booking.booking_status == 'CANCELLED':
            return error_response("This booking is already cancelled", 400)

        result = process_demo_cancellation_and_refund(db, booking, reason)
        return success_response(result, "Booking cancelled successfully. Simulated refund initiated.", 200)

    except Exception as e:
        db.rollback()
        return error_response(f"Cancellation failed: {str(e)}", 500)
    finally:
        db.close()

File: server/controllers/bookingController.js
const { query } = require('../config/db');
const { calculateAdvanceAmount } = require('../services/paymentService');
const { calculateCommission } = require('../services/commissionService');
const { processDemoRefund, calculateRefundAmount } = require('../services/refundService');

const generateBookingCode = () => {
  const randomNum = Math.floor(10000 + Math.random() * 90000);
  return `EH-BK-${randomNum}`;
};

const createBooking = async (req, res, next) => {
  try {
    const {
      vendorId,
      eventId,
      eventDate,
      serviceCategory,
      totalAmount,
      advancePercentage = 20
    } = req.body;

    if (!vendorId || !eventDate || !totalAmount) {
      return res.status(400).json({
        success: false,
        message: 'Vendor ID, event date, and total amount are required.'
      });
    }

    const userId = req.user ? req.user.id : 1;
    const bookingCode = generateBookingCode();

    // 1. Calculate Advance and Remaining
    const advanceInfo = calculateAdvanceAmount(totalAmount, advancePercentage);

    // 2. Calculate EventHub Commission (10%)
    const commissionInfo = calculateCommission(totalAmount, 10);

    // 3. Insert into Database
    const insertBookingSql = `
      INSERT INTO bookings (
        booking_code, user_id, event_id, vendor_id, event_date, service_category,
        total_amount, advance_percentage, advance_amount, remaining_amount,
        commission_rate, commission_amount
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `;

    const result = await query(insertBookingSql, [
      bookingCode,
      userId,
      eventId ? Number(eventId) : null,
      Number(vendorId),
      eventDate,
      serviceCategory || 'Event Service',
      advanceInfo.totalAmount,
      advanceInfo.advancePercentage,
      advanceInfo.advanceAmount,
      advanceInfo.remainingAmount,
      commissionInfo.commissionRate,
      commissionInfo.commissionAmount
    ]);

    const bookingId = result[0]?.insertId || 10;

    // Fetch vendor details for response
    const [vendors] = await query('SELECT * FROM vendors WHERE id = ?', [Number(vendorId)]);
    const vendor = vendors ? vendors[0] : null;

    res.status(201).json({
      success: true,
      message: 'Booking enquiry created successfully. Proceed to pay 20% advance to confirm.',
      booking: {
        id: bookingId,
        bookingCode,
        userId,
        eventId,
        vendorId: Number(vendorId),
        vendorName: vendor ? vendor.business_name : 'Vendor',
        vendorCategory: vendor ? vendor.category : serviceCategory,
        eventDate,
        totalAmount: advanceInfo.totalAmount,
        advancePercentage: advanceInfo.advancePercentage,
        advanceAmount: advanceInfo.advanceAmount,
        remainingAmount: advanceInfo.remainingAmount,
        commissionAmount: commissionInfo.commissionAmount,
        bookingStatus: 'PENDING',
        paymentStatus: 'PENDING'
      }
    });
  } catch (err) {
    next(err);
  }
};

const getBookingById = async (req, res, next) => {
  try {
    const bookingId = Number(req.params.id);
    const [bookings] = await query('SELECT * FROM bookings WHERE id = ?', [bookingId]);

    if (!bookings || bookings.length === 0) {
      return res.status(404).json({ success: false, message: 'Booking not found.' });
    }

    const booking = bookings[0];
    const [vendors] = await query('SELECT * FROM vendors WHERE id = ?', [booking.vendor_id]);
    const [payments] = await query('SELECT * FROM payments WHERE booking_id = ?', [bookingId]);
    const [refunds] = await query('SELECT * FROM refunds WHERE booking_id = ?', [bookingId]);

    res.json({
      success: true,
      booking: {
        ...booking,
        vendor: vendors ? vendors[0] : null,
        payments: payments || [],
        refund: refunds && refunds.length > 0 ? refunds[0] : null
      }
    });
  } catch (err) {
    next(err);
  }
};

const getUserBookings = async (req, res, next) => {
  try {
    const userId = req.user.id;
    const [bookings] = await query('SELECT * FROM bookings WHERE user_id = ? ORDER BY created_at DESC', [userId]);
    const [allVendors] = await query('SELECT * FROM vendors');
    const [allPayments] = await query('SELECT * FROM payments');
    const [allRefunds] = await query('SELECT * FROM refunds');

    const enriched = (bookings || []).map(b => {
      const vendor = allVendors.find(v => v.id === b.vendor_id);
      const payments = allPayments.filter(p => p.booking_id === b.id);
      const refund = allRefunds.find(r => r.booking_id === b.id);
      return {
        ...b,
        vendor,
        payments,
        refund
      };
    });

    res.json({
      success: true,
      count: enriched.length,
      bookings: enriched
    });
  } catch (err) {
    next(err);
  }
};

const getVendorBookings = async (req, res, next) => {
  try {
    const userId = req.user.id;
    // Find vendor associated with this user
    const [vendors] = await query('SELECT * FROM vendors WHERE user_id = ?', [userId]);
    const vendorId = vendors && vendors.length > 0 ? vendors[0].id : 1;

    const [bookings] = await query('SELECT * FROM bookings WHERE vendor_id = ? ORDER BY created_at DESC', [vendorId]);
    const [allUsers] = await query('SELECT id, name, email, phone FROM users');

    const enriched = (bookings || []).map(b => {
      const customer = allUsers.find(u => u.id === b.user_id);
      return {
        ...b,
        customerName: customer ? customer.name : 'Customer',
        customerPhone: customer ? customer.phone : '+91 98765 43210'
      };
    });

    res.json({
      success: true,
      vendor: vendors ? vendors[0] : null,
      bookings: enriched
    });
  } catch (err) {
    next(err);
  }
};

const updateBookingStatus = async (req, res, next) => {
  try {
    const bookingId = Number(req.params.id);
    const { status } = req.body; // 'CONFIRMED' or 'CANCELLED'

    if (!['CONFIRMED', 'CANCELLED', 'COMPLETED'].includes(status)) {
      return res.status(400).json({ success: false, message: 'Invalid booking status.' });
    }

    await query('UPDATE bookings SET booking_status = ? WHERE id = ?', [status, bookingId]);

    res.json({
      success: true,
      message: `Booking status updated to ${status}.`
    });
  } catch (err) {
    next(err);
  }
};

const cancelBooking = async (req, res, next) => {
  try {
    const bookingId = Number(req.params.id);
    const { reason = 'Customer requested cancellation', platformFee } = req.body;

    const refundResult = await processDemoRefund({
      bookingId,
      cancelledBy: req.user ? req.user.role : 'CUSTOMER',
      reason,
      platformFee
    });

    res.json({
      success: true,
      message: 'Booking cancelled and simulated refund initiated successfully.',
      refund: refundResult
    });
  } catch (err) {
    next(err);
  }
};

module.exports = {
  createBooking,
  getBookingById,
  getUserBookings,
  getVendorBookings,
  updateBookingStatus,
  cancelBooking
};
